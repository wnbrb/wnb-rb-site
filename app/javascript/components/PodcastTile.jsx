import React from 'react';
import PropTypes from 'prop-types';
import Card from './Card';
import Button from './Button';

import 'stylesheets/podcast-tile';

const PodcastTile = ({ podcasts }) => (
    <Card className="podcast-tile">
        {podcasts.map((podcast) => (
            <div className="podcast-listing" key={podcast.id}>
                <img src={podcast.tile_image} alt={podcast.tile_image_alt} />
                <a className="podcast-title" href={podcast.url}>
                    {podcast.title}
                </a>
                <Button type="white" className="min-w-[10rem]">
                    <a href={podcast.url} target="_blank" rel="noopener noreferrer">
                        Listen
                    </a>
                </Button>
            </div>
        ))}
    </Card>
);

PodcastTile.propTypes = {
    podcasts: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.number,
            tile_image: PropTypes.string,
            tile_image_alt: PropTypes.string,
            title: PropTypes.string,
            url: PropTypes.string,
        }),
    ),
};

export default PodcastTile;
