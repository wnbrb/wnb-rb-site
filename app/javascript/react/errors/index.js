export class UnauthorizedError extends Error {
    constructor(message) {
        super(message);
        this.name = 'UnauthorizedError';
    }
}
