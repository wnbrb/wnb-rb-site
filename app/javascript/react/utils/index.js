const MS_IN_DAY = 24 * 60 * 60 * 1000;

export const postedAtString = (createdAtString) => {
    const today = new Date();
    const postedDate = new Date(createdAtString);

    const diff = (+today - +postedDate) / MS_IN_DAY;

    if (diff <= 1) {
        return 'today';
    }

    if (diff < 7) {
        return `${Math.floor(diff)} days ago`;
    }

    if (diff < 30) {
        const weekDiff = Math.floor(diff / 7);
        const weekOrWeeks = weekDiff === 1 ? 'week' : 'weeks';

        return `${weekDiff} ${weekOrWeeks} ago`;
    }

    const monthDiff = Math.floor(diff / 30);
    const monthOrMonths = monthDiff === 1 ? 'month' : 'months';

    return `${monthDiff} ${monthOrMonths} ago`;
};
