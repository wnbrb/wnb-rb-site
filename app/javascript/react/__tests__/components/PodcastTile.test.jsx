/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react';
import { render } from '@testing-library/react';
import PodcastTile from '../../components/PodcastTile';

// Mock the Card and Button components to focus on PodcastTile logic
jest.mock('../../components/Card', () => ({ children, className }) => (
    <div className={className}>{children}</div>
));

jest.mock('../../components/Button', () => ({ children, type, className }) => (
    <button className={`${type} ${className}`}>{children}</button>
));

describe('PodcastTile Component', () => {
    const mockPodcasts = [
        {
            id: 1,
            tile_image: '/images/podcast1.jpg',
            tile_image_alt: 'Podcast 1 Cover',
            title: 'Amazing Tech Podcast',
            url: 'https://podcast1.com',
        },
        {
            id: 2,
            tile_image: '/images/podcast2.jpg',
            tile_image_alt: 'Podcast 2 Cover',
            title: 'Ruby Development Stories',
            url: 'https://podcast2.com',
        },
    ];

    const singlePodcast = [mockPodcasts[0]];

    // Snapshot tests
    it('should match snapshot with single podcast', () => {
        const { container } = render(<PodcastTile podcasts={singlePodcast} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot with multiple podcasts', () => {
        const { container } = render(<PodcastTile podcasts={mockPodcasts} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot with empty podcasts array', () => {
        const { container } = render(<PodcastTile podcasts={[]} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    // Functional tests
    it('should render all podcasts', () => {
        const { container } = render(<PodcastTile podcasts={mockPodcasts} />);
        const podcastListings = container.querySelectorAll('.podcast-listing');
        expect(podcastListings).toHaveLength(mockPodcasts.length);
    });

    it('should render podcast images with correct attributes', () => {
        const { container } = render(<PodcastTile podcasts={singlePodcast} />);
        const image = container.querySelector('img');

        expect(image).toHaveAttribute('src', singlePodcast[0].tile_image);
        expect(image).toHaveAttribute('alt', singlePodcast[0].tile_image_alt);
    });

    it('should render podcast titles as links', () => {
        const { getByText } = render(<PodcastTile podcasts={singlePodcast} />);
        const titleLink = getByText(singlePodcast[0].title);

        expect(titleLink.tagName).toBe('A');
        expect(titleLink).toHaveAttribute('href', singlePodcast[0].url);
        expect(titleLink).toHaveClass('podcast-title');
    });

    it('should render listen buttons with correct links', () => {
        const { container } = render(<PodcastTile podcasts={singlePodcast} />);
        const listenLink = container.querySelector('button a');

        expect(listenLink).toHaveAttribute('href', singlePodcast[0].url);
        expect(listenLink).toHaveAttribute('target', '_blank');
        expect(listenLink).toHaveAttribute('rel', 'noopener noreferrer');
        expect(listenLink).toHaveTextContent('Listen');
    });

    it('should use Card component with podcast-tile class', () => {
        const { container } = render(<PodcastTile podcasts={mockPodcasts} />);
        expect(container.firstChild).toHaveClass('podcast-tile');
    });

    it('should handle empty podcast array gracefully', () => {
        const { container } = render(<PodcastTile podcasts={[]} />);
        const podcastListings = container.querySelectorAll('.podcast-listing');
        expect(podcastListings).toHaveLength(0);
    });

    it('should render unique keys for each podcast', () => {
        const { container } = render(<PodcastTile podcasts={mockPodcasts} />);
        const podcastListings = container.querySelectorAll('.podcast-listing');

        // Each podcast listing should be present
        expect(podcastListings).toHaveLength(mockPodcasts.length);

        // Verify that both podcasts are rendered (check for unique content)
        expect(container).toHaveTextContent('Amazing Tech Podcast');
        expect(container).toHaveTextContent('Ruby Development Stories');
    });
});
