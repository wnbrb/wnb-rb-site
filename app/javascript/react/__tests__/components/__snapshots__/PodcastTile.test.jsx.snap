// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodcastTile Component should match snapshot with empty podcasts array 1`] = `
<div
  class="podcast-tile"
/>
`;

exports[`PodcastTile Component should match snapshot with multiple podcasts 1`] = `
<div
  class="podcast-tile"
>
  <div
    class="podcast-listing"
  >
    <img
      alt="Podcast 1 Cover"
      src="/images/podcast1.jpg"
    />
    <a
      class="podcast-title"
      href="https://podcast1.com"
    >
      Amazing Tech Podcast
    </a>
    <button
      class="white min-w-[10rem]"
    >
      <a
        href="https://podcast1.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        Listen
      </a>
    </button>
  </div>
  <div
    class="podcast-listing"
  >
    <img
      alt="Podcast 2 Cover"
      src="/images/podcast2.jpg"
    />
    <a
      class="podcast-title"
      href="https://podcast2.com"
    >
      Ruby Development Stories
    </a>
    <button
      class="white min-w-[10rem]"
    >
      <a
        href="https://podcast2.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        Listen
      </a>
    </button>
  </div>
</div>
`;

exports[`PodcastTile Component should match snapshot with single podcast 1`] = `
<div
  class="podcast-tile"
>
  <div
    class="podcast-listing"
  >
    <img
      alt="Podcast 1 Cover"
      src="/images/podcast1.jpg"
    />
    <a
      class="podcast-title"
      href="https://podcast1.com"
    >
      Amazing Tech Podcast
    </a>
    <button
      class="white min-w-[10rem]"
    >
      <a
        href="https://podcast1.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        Listen
      </a>
    </button>
  </div>
</div>
`;
